### ISSUE TITLE
Explicit listing of optional dependencies

### ISSUE THREAD

**Issue by ghisvail** [pydata/xarray#1879]
Could I get a list somewhere (in the README perhaps?) of all the optional dependencies for `xarray`, and whether they are recommended or just suggested. This would help me organize the Debian metadata for the new release.

For instance, given a feature with multiple backends, the dependencies for the preferred backend should be recommended, whilst the others should just be suggested.

Thanks,

**Comment by jhamman** [comment]
@ghisvail - Is http://xarray.pydata.org/en/latest/installing.html#optional-dependencies sufficient?

**Comment by ghisvail** [comment]
So, just to confirm, the recommended dependencies are just `dask`, `netCDF4` and `bottleneck`, right? The rest should be suggested.

What about `cyordereddict`? It is listed in the docs but its description says it is obsolete.

**Comment by maxim-lian** [comment]
> its description says it is obsolete.

obsolete on python 3.5+

**Comment by ghisvail** [comment]
And what about `toolz` which is used in `xarray/core/combine.py`?

**Comment by fmaussion** [comment]
yes, this list seems a bit outdated. rasterio should be mentioned too.

**Comment by ghisvail** [comment]
Indeed. `xarray/__init__.py` does import `auto_combine` from `xarray/core/combine.py`, so `toolz` should be listed as `install_depends` I guess.

**Comment by ghisvail** [comment]
And `open_rasterio` is also exposed by `xarray/__init__.py` and imports `rasterio` unconditionally. So again, that makes `rasterio` an `install_requires` too.

**Comment by fmaussion** [comment]
> open_rasterio is also exposed by xarray/__init__.py and imports rasterio unconditionally

I don't understand what you mean: as long as `open_rasterio` is not called by the user, `rasterio` doesn't need to be installed?

**Comment by fmaussion** [comment]
arguably, we could catch the import error and send a more informative error message

**Comment by ghisvail** [comment]
Well, I am considering (perhaps wrongly) what's in `xarray/__init__.py` as the public API of `xarray`. With that logic in mind, any unconditional import should be treated as an install dependency. Otherwise, users will indeed get an `ImportError` which is not very good.

Back to the original topic, should `toolz` and `rasterio` be strict dependencies, recommended or suggested then?

**Comment by fmaussion** [comment]
rasterio is only required if you want to open a specific type of files (geotiffs): the vast majority of our users won't touch `open_rasterio`. I guess that by your definition it would be suggested? Is rasterio going to make it to debian as an apt package?

**Comment by ghisvail** [comment]
> I guess that by your definition it would be suggested?

Correct.

> Is rasterio going to make it to debian as an apt package?

That's already the case, see: https://packages.debian.org/source/sid/rasterio

What about `toolz` used by `auto_combine`?

**Comment by shoyer** [comment]
> What about toolz used by auto_combine?

`toolz` should probably be listed explicitly as an optional dependencies, but I didn't bother since it's a hard dependency of `dask.array`, and we only use `toolz` in cases where `dask.array` is also required. (In theory, dask.array could drop its use of toolz internally, but that's unlikely to happen anytime soon.)

>  For instance, given a feature with multiple backends, the dependencies for the preferred backend should be recommended, whilst the others should just be suggested.

I don't think the difference between recommended and suggested dependencies is relevant for xarray. We don't have any features with multiple backends where a single backend is strictly preferred.
